### PR TITLE
perf: modify YCSB for LDR testing

### PR DESCRIPTION
**Pull request by ajstorm**
For testing of logical data replication, there is a need to create a write only workload with a schema that contains several columns (i.e. not kv, which has only two). This can easily be tucked into ycsb with an internal-only workload "G". I'd understand if folks didn't want to muddy the ycsb waters with this workload modification, but doing it this way allows us to make faster progress on the LDR side, so I figured I'd give it a try.

Epic: CRDB-38996
Release note: None

### CODE REVIEW

**Comment by nvanbenschoten** [comment]
> This can easily be tucked into ycsb with an internal-only workload "G". I'd understand if folks didn't want to muddy the ycsb waters with this workload modification, but doing it this way allows us to make faster progress on the LDR side, so I figured I'd give it a try.

Yeah, I personally don't love the approach of defining a new "workload G" for this reason. YCSB is a fairly well defined workload (ref: https://courses.cs.duke.edu/fall13/cps296.4/838-CloudPapers/ycsb.pdf) and it would be a shame to start deviating from the spec at this level. That said, the official YCSB workload is even more flexible, allowing the request ratio to be fully [customized](https://github.com/brianfrankcooper/YCSB/blob/master/workloads/workload_template) in custom workloads.

Instead of adding a "workload G", we could add flags for each of `readFreq`, `insertFreq`, `updateFreq`, `scanFreq`, and `readModifyWriteFreq`. This would allow users to make their own workloads instead of using a predefined workload.

**Comment by srosenberg** [comment]
> Instead of adding a "workload G", we could add flags for each of `readFreq`, `insertFreq`, `updateFreq`, `scanFreq`, and `readModifyWriteFreq`. This would allow users to make their own workloads instead of using a predefined workload.

I'd also prefer this option; this would give us `--workload=custom`.

**Comment by ajstorm** [comment]
Fair enough, I went for the quick fix.  PTAL, as I've implemented the desired fix.  

Of note however, is that we need one additional fix for LDR testing.  Currently the kit doesn't apply the seed when generating insert keys.  As a result, regardless of the seed provided, it will always generate the same sequence of keys for insertin.  This causes problems for LDR as two runs initiated on separate clusters, will generate a stream of conflicts when replicated.  The change uses the seed to salt the key before the hashing occurs, which generates a distinct set of keys for different seed values.

**Review by nvb — COMMENTED** [review]
> Currently the kit doesn't apply the seed when generating insert keys. As a result, regardless of the seed provided, it will always generate the same sequence of keys for insertin. This causes problems for LDR as two runs initiated on separate clusters, will generate a stream of conflicts when replicated. The change uses the seed to salt the key before the hashing occurs, which generates a distinct set of keys for different seed values.

Won't this break all other uses of ycsb, which rely on zipfian request distribution over a known set of keys? For example, in YCSB-C, we insert 1,000,000 keys during an `workload init` phase, and then perform 100% reads over those keys in a `workload run` phase.

Can we accomplish the same task by using the `--insert-start` and `--max-ops` flags? These were added in https://github.com/cockroachdb/cockroach/commit/dfc1e572f83338a9fb7f3d5c1fba79cdfc56df35, which even references "cases where multiple clients are running workloads in parallel".

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125001#-:-NznjI6HDuinLcxfIz4X:bn1g066)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @herkolategan, and @srosenberg)

___
*[`pkg/workload/ycsb/ycsb.go` line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/125001#-Nznix8PEe1rPjtZhPW7:-Nznix8Q8b0arwKLuok6:b-9r7gt6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ef4e5dbbc4d261fa51c33f3750b89fb646acadb9/pkg/workload/ycsb/ycsb.go#L154)):*
> ```Go
> 		g.flags.Uint64Var(&g.minScanLength, `min-scan-length`, 1, `The minimum length for scan operations. Primarily used for workload E.`)
> 		g.flags.Uint64Var(&g.maxScanLength, `max-scan-length`, 1000, `The maximum length for scan operations. Primarily used for workload E.`)
> 		g.flags.BoolVar(&g.customWorkload, `custom-workload`, false, `Allows the specification of a custom workload in conjunction with the *-freq flags. (default false)`)
> ```

Instead of adding this flag, could we just accept `--workload=custom`? That's what @srosenberg was suggesting, and it seems nice.

We could also assert that all of the following distribution flags are 0 if not custom, and add up to 1.0 if custom.

**Comment by ajstorm** [comment]
> Won't this break all other uses of ycsb, which rely on zipfian request distribution over a known set of keys? For example, in YCSB-C, we insert 1,000,000 keys during an workload init phase, and then perform 100% reads over those keys in a workload run phase.

It shouldn't, no.  The changes I've made are all within the `hashKey` function, which accepts the sequential key that is used in both the read and write paths.  In the zipfian case, we determine an internal key to use [first](https://github.com/cockroachdb/cockroach/pull/125001/files#diff-f3aa9f413e52eca7d64bf33c9493ec426a0c54aa4dca7a9d948721aa365e96c0L669), and then pass it to the hash function to determine the external key (i.e. the one in the pk column of the table).



> Can we accomplish the same task by using the --insert-start and --max-ops flags? These were added in https://github.com/cockroachdb/cockroach/commit/dfc1e572f83338a9fb7f3d5c1fba79cdfc56df35, which even references "cases where multiple clients are running workloads in parallel".

I didn't realize that using the `--insert-start` flag was a possibility (partially because I'm not that familiar with YCSB, and partially because the help for that flag seems to imply - to me at least - that it's only usable in the sequential insert case).  

We definitely could go down that path, but the reasons I coded it up this way are:

- it seemed odd to me that only the read path respects the seed
- the `--insert-start` flag seems more fragile (if you set it incorrectly, when you get to a point where two clusters overlap in the keyspace, the kit will produce a never-ending stream of conflicts.  

It seems less fragile to have a 64-bit key as the seed as:

1. It's far less likely that two clusters will randomly collide when setting the seed
2. You don't have to worry about a continuous collision situation

Thoughts?

> Instead of adding this flag, could we just accept --workload=custom? That's what @srosenberg was suggesting, and it seems nice.

Yeah.  Sorry, missed that in the first review.

> We could also assert that all of the following distribution flags are 0 if not custom, and add up to 1.0 if custom.

Will add.

**Comment by nvanbenschoten** [comment]
> It shouldn't, no. The changes I've made are all within the hashKey function, which accepts the sequential key that is used in both the read and write paths. In the zipfian case, we determine an internal key to use [first](https://github.com/cockroachdb/cockroach/pull/125001/files#diff-f3aa9f413e52eca7d64bf33c9493ec426a0c54aa4dca7a9d948721aa365e96c0L669), and then pass it to the hash function to determine the external key (i.e. the one in the pk column of the table).

Isn't the effect of this change that the internal-to-external key mapping (`hashKey`) changes on each run of YCSB? So if we run `workload init ycsb` and then later `workload run ycsb`, are we going to read/update any of the rows that were imported during the `init` call?

> the --insert-start flag seems more fragile (if you set it incorrectly, when you get to a point where two clusters overlap in the keyspace, the kit will produce a never-ending stream of conflicts.

On the other hand, it gives correct users a guarantee of no conflicts, instead of leaving it up to random change.

**Comment by ajstorm** [comment]
>Isn't the effect of this change that the internal-to-external key mapping (hashKey) changes on each run of YCSB? So if we run workload init ycsb and then later workload run ycsb, are we going to read/update any of the rows that were imported during the init call?

As long as the same seed is provided for both executions (init and run) then the internal-to-external mapping should be consistent.  The change consistently salts based on the provided seed.  This need is also called out in the pre-existing help message, which is what led me to believe that the seed would be applied on both init and run.  

```
      --seed uint                           Random seed. Must be the same in 'init' and 'run'. Default changes in each run (default 6912887537232448832)
```

At the bare minimum, this help message is wrong, but my aim to correct it is also what motivated the proposed change.

>On the other hand, it gives correct users a guarantee of no conflicts, instead of leaving it up to random change.

Aren't we dealing with random chance in both cases?  Couldn't the hash function could produce collisions in the same way that the salting via the seed can?

**Comment by nvanbenschoten** [comment]
> As long as the same seed is provided for both executions (init and run) then the internal-to-external mapping should be consistent. The change consistently salts based on the provided seed. This need is also called out in the pre-existing help message, which is what led me to believe that the seed would be applied on both init and run.

Interesting, nice spot. It looks like this `--seed` flag was only recently randomized on each run in https://github.com/cockroachdb/cockroach/commit/7b71595569170393e2c2e94bee5d8c7c0b440019, whereas before it defaulted to `1`.

We don't seem to use the flag very commonly, so mandating it will be a breaking change. For instance, [this is a roachtest](https://github.com/cockroachdb/cockroach/blob/c7f37e813dbd1e6d2afdb88c5e9cd986dc061299/pkg/cmd/roachtest/tests/split.go#L104) that would break if `workload init ycsb` and `workload run ycsb` don't create and then use the same dataset. I think the main ycsb roachtest would be ok because it uses the `workload run --init` shorthand, but that's only coincidental. Manual testing would also break, like that presented [in our docs here](https://www.cockroachlabs.com/docs/stable/cockroach-workload#run-the-ycsb-workload). What's worse is that it would break silently, with most UPDATEs becoming no-ops and performance improving.

There are 69 uses of `workload init` in the cockroachdb code base and not a single one uses `--seed`. Are these all now incorrectly operating over an empty dataset? I think probably not, but only because `--seed` happens to not influence the row identities for most workload generators.

@renatolabs we may want to rethink that commit, which lifted the "Must be the same in 'init' and 'run'" requirement from a single workload (`rand`) up to every workload. `workload init` followed by `workload run` is a very common pattern in roachtests, in our manual testing, and [in our docs](https://www.cockroachlabs.com/docs/stable/cockroach-workload). Are we now saying that all of these uses that don't pass a `--seed` flag to both are broken? If so, should we require the `--seed` flag be passed?

@ajstorm coming back to this PR, I think your change here to salt the key with the seed would actually be ok _if we went back to a constant default seed_. This would allow the common case of not passing a `--seed` to work properly, while LDR could use a different `--seed` on each cluster.

For now though, I would still suggest that LDR just uses `--insert-start` to get unblocked. This is how we usually use YCSB when running multiple clients in parallel. It's also how the official YCSB recommends that you run multiple clients in parallel: https://github.com/brianfrankcooper/YCSB/wiki/Running-a-Workload-in-Parallel.

> Aren't we dealing with random chance in both cases? Couldn't the hash function could produce collisions in the same way that the salting via the seed can?

Yeah, I think you're correct about this. We're playing dice no matter what we do.

File: pkg/workload/ycsb/ycsb.go [hunk 1]
```diff
@@ -619,7 +646,13 @@ const (
 )
 
 func (yw *ycsbWorker) hashKey(key uint64) uint64 {
-	yw.hashBuf = [8]byte{} // clear hashBuf
+	yw.hashBuf = [20]byte{} // clear hashBuf
```
Comment (nvanbenschoten) [inline pkg/workload/ycsb/ycsb.go:649]: Should this be `binary.MaxVarintLen64`?

Comment (ajstorm) [reply pkg/workload/ycsb/ycsb.go]: Fixed

**Comment by ajstorm** [comment]
> I think your change here to salt the key with the seed would actually be ok if we went back to a constant default seed. This would allow the common case of not passing a --seed to work properly, while LDR could use a different --seed on each cluster.
For now though, I would still suggest that LDR just uses --insert-start to get unblocked. This is how we usually use YCSB when running multiple clients in parallel. It's also how the official YCSB recommends that you run multiple clients in parallel: [Wiki: Running a Workload in Parallel (brianfrankcooper/YCSB)](https://github.com/brianfrankcooper/YCSB/wiki/Running-a-Workload-in-Parallel).

I think I'm stuck waiting for @renatolabs to chime in.  There's one remaining problem with using the --insert-start flag - it's defined as an int.  Even with a relatively modest 20K inserts/second, we'll blow through that total in a one day continuous run.  I thought it would be as simple thing to change that to an int64 type, but the int type is plumbed deep throughout workload, and the refactor will take some time.  I'd rather not have to go down that path if we're going to revert Renato's PR anyway, and the salting approach will work.  

Actually, come to think of it, will the salting approach make anything _worse_ in the interim?  Aren't we currently stuck with the need to pass the same seed to both init and run?  I appreciate that the current tests are likely broken, but salting shouldn't break them any more, and should eventually work when they're fixed.  

Is there any harm in merging the salting approach now while we wait for Renato's opinion on [7b71595](https://github.com/cockroachdb/cockroach/commit/7b71595569170393e2c2e94bee5d8c7c0b440019)?

**Comment by nvanbenschoten** [comment]
> There's one remaining problem with using the --insert-start flag - it's defined as an int. Even with a relatively modest 20K inserts/second, we'll blow through that total in a one day continuous run. I thought it would be as simple thing to change that to an int64 type, but the int type is plumbed deep throughout workload, and the refactor will take some time.

The width of an `int` in Go is platform dependent (https://go.dev/ref/spec#Numeric_types). On a 64-bit system, an `int` and an `int64` have the same size. Since we almost always run on 64-bit systems, we shouldn't need to change the flag to an `int64`. With a relatively modest 20K inserts/second, we'll have `(2^64 / 20k ins/sec) * (1 / (60*60*24) day/sec) = 10B` days before we run into issues.

> Actually, come to think of it, will the salting approach make anything worse in the interim? Aren't we currently stuck with the need to pass the same seed to both init and run? I appreciate that the current tests are likely broken, but salting shouldn't break them any more, and should eventually work when they're fixed.

The thing is that things are not actually broken yet, because we don't require `init` and `run` to have the same `--seed` flag for ycsb. It's just the help text which is incorrect. We will only break things if we start using the random seed to decide the keys in the dataset, because then `init` will import a different dataset than the one `run` will use.

**Comment by ajstorm** [comment]
>The width of an int in Go is platform dependent ([go.dev/ref/spec#Numeric_types](https://go.dev/ref/spec#Numeric_types)). On a 64-bit system, an int and an int64 have the same size. Since we almost always run on 64-bit systems, we shouldn't need to change the flag to an int64. With a relatively modest 20K inserts/second, we'll have (2^64 / 20k ins/sec) * (1 / (60*60*24) day/sec) = 10B days before we run into issues.

Right, of course.  This is the same in C/C++ but my memory of the `int` type is foggy, as in my old job, using it was banned since we started supporting 64-bit platforms, because it couldn't be reliably reasoned about.  Instead, we enforced use of explicit int32/int64.

I think the PR is where you wanted it, and is ready for another look.

**Review by nvb — APPROVED** [review]
:lgtm: thanks for the discussion on this. Hopefully LDR testing is now unblocked.

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/125001#-:-O--VX-Y-MSdbS1QNpHf:b6abvnw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @ajstorm, @herkolategan, and @srosenberg)

___
*[`-- commits` line 7 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/125001#-O--V9id5EU8Esp82zGw):*
nit: we can remove the "The second change salts ..." sentence now.

**Comment by nvanbenschoten** [comment]
> Indeed, although I would just say that the help message is misleading, as already pointed out here. Needless to say, the need for the same seed for init and run only exists for workloads where there is a coupling between init time and run time which, in my understanding, is not the case for most workloads.

Yes, exactly. Right now the help message makes it sound like that is the case for all workloads, despite that not being the case.

**Pull request merged**